pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -291,15 +465,64 @@ pub mod pallet {
             Ok(reanchored)
         }
 
+        fn clean_up_expired_in_flight_orders(n: BlockNumberFor<T>) {
+            let mut ttl_queue = InFlightOrdersTtl::<T>::get();
+            let mut maybe_oldest_entry;
+            let mut oldest_block_number;
+            let mut query_id;
+
+            let mut query_ids_to_remove = vec![];
+
+            loop {
+                maybe_oldest_entry = ttl_queue.first().copied();
+                (oldest_block_number, query_id) = if let Some(oldest_entry) = maybe_oldest_entry {
+                    oldest_entry
+                } else {
+                    break;
+                };
+
+                // No entry at current block number found
+                if oldest_block_number > n {
+                    break;
+                }
+
+                ttl_queue.remove(&(oldest_block_number, query_id));
+
+                query_ids_to_remove.push(query_id);
+            }
+
+            // Return early if there is nothing to remove.
+            if query_ids_to_remove.is_empty() {
+                return;
+            }
+
+            let mut query_to_para_mapping = QueryIdToParaId::<T>::get();
+            let mut in_flight_orders = InFlightOrders::<T>::get();
+
+            for query_id_to_remove in query_ids_to_remove {
+                let para_id = query_to_para_mapping.remove(&query_id_to_remove).expect("If an entry exists in InFlightOrdersTtl then \
+                it must exists on QueryIdToParaId mapping, if not we have storage inconsistency and better to crash; qed.");
+                in_flight_orders.remove(&para_id);
+            }
+
+            InFlightOrdersTtl::<T>::put(ttl_queue);
+            InFlightOrders::<T>::put(in_flight_orders);
+            QueryIdToParaId::<T>::put(query_to_para_mapping);
+        }
+
         /// Send an XCM message to the relay chain to try to buy a core for this para_id.
         fn on_collator_instantaneous_core_requested(para_id: ParaId) -> DispatchResult {
+            Self::clean_up_expired_in_flight_orders(<frame_system::Pallet<T>>::block_number());
+
+            let pending_blocks = PendingBlocks::<T>::get();
+            if pending_blocks.contains(&para_id) {
+                return Err(Error::<T>::BlockProductionPending.into());
+            }
```

Comment:
This looks wrong, because when an xcm message is successful and the core is bought, the collator can still fail to produce a block. In that scenario we want to allow another collator to try to buy a core again after the timeout.

And also we cannot prevent a third party from manually buying a core in the relay chain, using their own account, so it is possible that a parathread produces a block when we haven't sent an xcm message.

So I would remove the `PendingBlocks` storage item.